pkg/resources/task.go

```diff
@@ -0,0 +1,645 @@
+		return err
+	}
+
+	err = data.Set("warehouse", t.Warehouse)
+	if err != nil {
+		return err
+	}
+
+	err = data.Set("schedule", t.Schedule)
+	if err != nil {
+		return err
+	}
+
+	err = data.Set("comment", t.Comment)
+	if err != nil {
+		return err
+	}
+
+	if t.Predecessors != nil {
+		err = data.Set("after", t.GetPredecessorName())
+		if err != nil {
+			return err
+		}
+	}
+
+	err = data.Set("when", t.Condition)
+	if err != nil {
+		return err
+	}
+
+	err = data.Set("sql_statement", t.Definition)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// CreateTask implements schema.CreateFunc
+func CreateTask(data *schema.ResourceData, meta interface{}) error {
+	var (
+		err       error
+		db        = meta.(*sql.DB)
+		database  = data.Get("database").(string)
+		dbSchema  = data.Get("schema").(string)
+		name      = data.Get("name").(string)
+		warehouse = data.Get("warehouse").(string)
+		sql       = data.Get("sql_statement").(string)
+		enabled   = data.Get("enabled").(bool)
+	)
+
+	builder := snowflake.Task(name, database, dbSchema)
+	builder.WithWarehouse(warehouse)
+	builder.WithStatement(sql)
+
+	// Set optionals
+	if v, ok := data.GetOk("schedule"); ok {
+		builder.WithSchedule(v.(string))
+	}
+
+	if v, ok := data.GetOk("session_parameters"); ok {
+		builder.WithSessionParameters(v.(map[string]interface{}))
+	}
+
+	if v, ok := data.GetOk("user_task_timeout_ms"); ok {
+		builder.WithTimeout(v.(int))
+	}
+
+	if v, ok := data.GetOk("comment"); ok {
+		builder.WithComment(v.(string))
+	}
+
+	if v, ok := data.GetOk("after"); ok {
+		root, err := getRootTaskAndSuspend(data, meta)
+		if err != nil {
+			return err
+		}
+		defer resumeTask(root, meta)
+
+		builder.WithDependency(v.(string))
+	}
+
+	if v, ok := data.GetOk("when"); ok {
+		builder.WithCondition(v.(string))
+	}
+
+	q := builder.Create()
+	err = snowflake.Exec(db, q)
+	if err != nil {
+		return errors.Wrapf(err, "error creating task %v", name)
+	}
+
+	if enabled {
+		q = builder.Resume()
+		err = snowflake.Exec(db, q)
+		if err != nil {
+			return errors.Wrapf(err, "error starting task %v", name)
+		}
+	}
+
+	taskID := &taskID{
+		DatabaseName: database,
+		SchemaName:   dbSchema,
+		TaskName:     name,
+	}
+	dataIDInput, err := taskID.String()
+	if err != nil {
+		return err
+	}
+	data.SetId(dataIDInput)
+
+	return ReadTask(data, meta)
+}
+
+// UpdateTask implements schema.UpdateFunc
+func UpdateTask(data *schema.ResourceData, meta interface{}) error {
+	// https://www.terraform.io/docs/extend/writing-custom-providers.html#error-handling-amp-partial-state
+	data.Partial(true)
+
+	taskID, err := taskIDFromString(data.Id())
+	if err != nil {
+		return err
+	}
+
+	var (
+		db       = meta.(*sql.DB)
+		database = taskID.DatabaseName
+		dbSchema = taskID.SchemaName
+		name     = taskID.TaskName
+		builder  = snowflake.Task(name, database, dbSchema)
+	)
+
+	root, err := getRootTaskAndSuspend(data, meta)
+	if err != nil {
+		return err
+	}
+	defer resumeTask(root, meta)
```

Comment:
@henriblancke you need to handle the case where the root was suspended

Reply:
@chriskuchin check `getRootTask` and `getRootTaskAndSuspend`. I added some logic there to not return the root task builder when the root task is already suspended that way the logic here and in other places stays concise and simple? Let me know if you think it is misleading or makes the logic hard to follow.

Reply:
ahh nice I think though you are missing an edge case where you change multiple things including the state. i.e. i think if you were to alter some field plus change enabled -> false on a root node you would end up with enabled true at the end maybe?

Reply:
@chriskuchin how would you handle that? It depends in what order terraform executes changes and how it builds its dag? If there is some order through tf variables, it would most likely alter the root task first before going onwards to the children or am I thinking this all the wrong way?

Reply:
you get a single nodes updates all at once. So if the node you are modifying is the root and currently enabled You will suspend it then if there is a change to suspend the node you are on further down the method call you will go suspend it. Then your defer will trigger and reenable it

Reply:
The whole state machine around this took me forever and I really didn't love my code to handle it but my acceptance tests should exercise this case i believe

Reply:
I think you may be able to solve the problem by if you handle a change in state you set the root to nil when you handle the node state change so your resume is a no op

Reply:
@chriskuchin added your acceptance tests

Reply:
@chriskuchin I want to understand better how you handled this edge case and how you would prevent things from being executed in parallel? I think by default terraform executes 10 operations in parallel as it walks the graph. If there is no relation that helps tf define the order when it builds the graph, there is no way to prevent it from running both operations in your edge case at the same time? Let me know if I'm looking at this in the wrong way

Reply:
could people use `depends_on` to avoid this edge case?

Reply:
There are 2 edge cases 1 that i don't think you can handle which is you concurrently modify 2 nodes, but can be prevented by using references from tasks for the after clause. and the second that is if you disable an enabled task on a node that is the root node then you will reactivate it with your defer

Reply:
the operations from my testing happen on a node basis not a change in node basis. So you get all of a nodes changes in one call to the update method

Reply:
depends_on or using variable references would fix the first concurrent execution problem but doesn't fix the other. Have you run my acceptance tests against your code?

Reply:
Also if it is easier I am happy to hop on a zoom for a quick conversation

Reply:
@chriskuchin lets hop on [a zoom call](https://us04web.zoom.us/j/75028769614?pwd=b2JPZWF0UXBUN1pGTE5Gd1Q4UmtmUT09)